lib/vnet/customdnszonevalidator.go

```diff
@@ -0,0 +1,71 @@
+package vnet
+
+import (
+	"context"
+	"log/slog"
+	"net"
+	"slices"
+	"sync"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	clusterTXTRecordPrefix = "teleport-cluster="
+)
+
+type lookupTXTFunc = func(ctx context.Context, domain string) (txtRecords []string, err error)
+
+// customDNSZoneValidator validates custom DNS zones by making sure that they have a DNS TXT record of
+// "teleport-cluster=<cluster-name>" for the cluster in which they are used. This is meant to avoid the
+// possibility of a rogue application advertising a public_addr with a DNS name not controlled by the cluster
+// admins, which could be used to trick VNet users.
+//
+// After finding that a zone is valid once, this is cached indefinitely. Invalid zones are a misconfiguration
+// so we don't cache negative results
+type customDNSZoneValidator struct {
+	lookupTXT  lookupTXTFunc
+	validZones map[string]struct{}
+	mu         sync.RWMutex
+}
+
+func newCustomDNSZoneValidator(lookupTXT lookupTXTFunc) *customDNSZoneValidator {
+	if lookupTXT == nil {
+		var resolver net.Resolver
+		lookupTXT = resolver.LookupTXT
+	}
+	return &customDNSZoneValidator{
+		lookupTXT:  lookupTXT,
+		validZones: make(map[string]struct{}),
+	}
+}
+
+// validate returns an error if [customDNSZone] is not valid for [clusterName].
+func (c *customDNSZoneValidator) validate(ctx context.Context, clusterName, customDNSZone string) error {
+	c.mu.RLock()
+	_, ok := c.validZones[customDNSZone]
+	c.mu.RUnlock()
+	if ok {
+		return nil
+	}
+
+	requiredTXTRecord := clusterTXTRecordPrefix + clusterName
+	slog.InfoContext(ctx, "Checking validity of custom DNS zone by querying for required TXT record.", "zone", customDNSZone, "record", requiredTXTRecord)
+
+	records, err := c.lookupTXT(ctx, customDNSZone)
+	if err != nil {
+		return trace.Wrap(err, "looking up TXT records for %q", customDNSZone)
```

Comment:
I'm in the process of setting up a cluster so that I can fully test custom DNS zones. I wanted to check what happens when the TXT records are not set up and this is the error I got:

```
2024-05-31T16:47:07+02:00 DEBU [VNET]      Error handling UDP conn. request:{53 fd25:36e1:3b77::2 53419 fd25:36e1:3b77::1} error:[
ERROR REPORT:
Original Error: *net.DNSError lookup <my custom DNS zone> on 1.1.1.1:53: no such host
Stack Trace:
	github.com/gravitational/teleport/lib/vnet/customdnszonevalidator.go:57 github.com/gravitational/teleport/lib/vnet.(*customDNSZoneValidator).validate
	github.com/gravitational/teleport/lib/vnet/app_resolver.go:200 github.com/gravitational/teleport/lib/vnet.(*TCPAppResolver).fqdnMatchesProfile
	github.com/gravitational/teleport/lib/vnet/app_resolver.go:150 github.com/gravitational/teleport/lib/vnet.(*TCPAppResolver).ResolveTCPHandler
	github.com/gravitational/teleport/lib/vnet/vnet.go:568 github.com/gravitational/teleport/lib/vnet.(*Manager).ResolveA.func1
	golang.org/x/sync@v0.7.0/singleflight/singleflight.go:198 golang.org/x/sync/singleflight.(*Group).doCall.func2
	golang.org/x/sync@v0.7.0/singleflight/singleflight.go:200 golang.org/x/sync/singleflight.(*Group).doCall
	golang.org/x/sync@v0.7.0/singleflight/singleflight.go:113 golang.org/x/sync/singleflight.(*Group).Do
	github.com/gravitational/teleport/lib/vnet/vnet.go:559 github.com/gravitational/teleport/lib/vnet.(*Manager).ResolveA
	github.com/gravitational/teleport/lib/vnet/dns/dns.go:202 github.com/gravitational/teleport/lib/vnet/dns.(*Server).handleDNSMessage
	github.com/gravitational/teleport/lib/vnet/dns/dns.go:130 github.com/gravitational/teleport/lib/vnet/dns.(*Server).HandleUDP
	github.com/gravitational/teleport/lib/vnet/vnet.go:528 github.com/gravitational/teleport/lib/vnet.(*Manager).handleUDPConcurrent
	github.com/gravitational/teleport/lib/vnet/vnet.go:479 github.com/gravitational/teleport/lib/vnet.(*Manager).handleUDP.func1
	runtime/asm_arm64.s:1222 runtime.goexit
User Message: resolving A request for &#34;tcp-postgres.<my custom DNS zone>.&#34;
	resolving TCP handler for fqdn &#34;tcp-postgres.<my custom DNS zone>.&#34;
		validating custom DNS zone &#34;<my custom DNS zone>&#34; for cluster &#34;<my cluster>&#34;
			looking up TXT records for &#34;<my custom DNS zone>&#34;
				lookup <my custom DNS zone> on 1.1.1.1:53: no such host] vnet/vnet.go:529
2024-05-31T16:47:07+02:00 DEBU [VNET]      Finished handling UDP request. request:{53 fd25:36e1:3b77::2 53419 fd25:36e1:3b77::1} vnet/vnet.go:531
2024-05-31T16:47:07+02:00 DEBU [VNET]      Got HUP or ERR, canceling request context and closing UDP conn. request:{53 fd25:36e1:3b77::2 53419 fd25:36e1:3b77::1} vnet/vnet.go:520
```

The error says "no such host". After some googling, I found https://github.com/golang/go/issues/65288 which says that `net.LookupTXT` returns "no such host" when there are no TXT records.

Could we maybe extend the error message here with something like `are VNet TXT records set on <custom DNS zone?`